hostchk.1
.\" section 1 man page for hostchk.sh
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec misfeatures and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH hostchk.sh 1 "27 October 2022" "hostchk.sh" "IOCCC tools"
.SH NAME
hostchk.sh \- run a series of checks on your system to help determine if you can correctly use the mkiocccentry toolkit
.SH SYNOPSIS
\fBhostchk.sh\fP [\-h] [\-V] [\-v level] [\-D dbg_level] [\-c cc] [\-w work_dir] [\-f]
.SH DESCRIPTION
\fBhostchk.sh\fP runs a series of checks to determine if you will be able to create an acceptable IOCCC entry tarball.
If any step fails we recommend you use the \fBbug_report(1)\fP script to create a log file for filing a bug.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB\-v \fIlevel\fP\fP
Set verbosity level to \fIlevel\fP.
.TP
\fB\-D \fIlevel\fP\fP
Set verbosity level for the tests the script runs to \fIlevel\fP.
.TP
\fB\-c \fIcc\fP\fP
Set path to the compiler (cc) to \fIcc\fP.
.TP
\fB\-w \fIwork_dir\fP\fP
Change work directory to\fIwork_dir\fP.
.TP
\fB\-f\fP
Fast compile mode.
Instead of testing one required system header file at a time it does it all at once.
The default is not to use \fB\-f\fP because it will give more information to help identify the problem (and also let you know which files work).
.SH EXIT STATUS
If no issues are detected the script will exit 0.
Otherwise it will exit non\-zero.
.SH NOTES
.PP
Even if the script exits 0 it does not necessarily mean there is no problem.
If you feel there is a problem that was not detected you may still open an issue on the GitHub page (or email the Judges).
In that case you should use \fBbug_report(1)\fP and include the log that it creates in your report.
.PP
The output can seem redundant if it fails and \fBhostchk(1)\fP was run via the Makefile since it will issue a warning that will tell you to use
\fBhostchk(1)\fP and also tell you to use \fBbug_report(1)\fP which itself will tell you to use \fBhostchk(1)\fP which will trigger the warning again.
Sorry in advance if (to use an American English phrase which I will happily use for a pun for a pun not made is a wasted opportunity) this throws you for a loop :\-) but we feel that it's better that the warning is really stressed as it most likely means that you won't be able to use the toolkit to successfully create a valid IOCCC tarball.
But now that you're in the loop it shouldn't even matter! :\-)
.SH BUGS
.PP
Right now there are no known bugs except that it is currently incomplete.
Nevertheless, if you feel there is an issue with this tool you may open an issue at the GitHub issues page.
Depending on the issue it might be a good idea to run \fBbug_report(1)\fP and attach the log file it creates.
.PP
Please remember that just because the script reports issues does not mean that there's a bug with the script itself.
Instead it probably is a problem with one of the tools on your system which the script detected and is reporting.
If that is the case please do use \fBbug_report(1)\fP and attach the log file it creates in your report.
\fBbug_report(1)\fP will tell you how and where to report bugs.
.SH EXAMPLES
.PP
.nf
Run script in fast mode:

.RS
\fB
 ./hostchk.sh \-f\fP
.fi
.RE
.SH SEE ALSO
\fBbug_report(1)\fP.


CHANGES.md
# Major changes to the IOCCC entry toolkit

## Release 0.7 2022-11-03

Released IOCCC entry toolkit v0.7 2022-11-03

`chkentry` is now code complete!
Added test suite for `chkentry`. More test files will be added in time.

Removed prefix in bison and flex code. The programmer's apology and warning are
still in `sorry.tm.ca.h`.

A test suite for `txzchk` was added.

Removed the `MAX_DIR_KSIZE` limit. There still is a maximum tarball size but now
there also is a limit of the number of files in the tarball. `txzchk` and
`mkiocccentry` were updated for this.

There probably were other changes as well but we're one step closer to IOCCCMOCK
with `chkentry` being complete!


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim
this is *NOT* complete. :-)


## Release 0.5 2022-07-10

The `jsemtblgen` tool is code complete.

Added `is_reserved()` function to `rule_count.c`.

Changed the Official iocccsize version from 28.10 2022-03-15 to 28.11
2022-07-10. Updated `test_JSON` tree and `json_teststr.txt` to reflect the new
`IOCCCSIZE_VERSION` value.

Added and improved a number of man pages.

Added `Makefile` rule to build HTML files from all man pages.

Both the `parse_json()` and `parse_json_file()` top level functions of the JSON
parser return a JSON node parse tree when the JSON is valid.

Trimmed the JSON parser to not include code that does not strictly belong to the
generic JSON parser.  Removed dead C code.

Updated dbg facility to version 2.4 2022-06-26.  Version 2.4 2022-06-26 contains

hostchk.sh
#!/usr/bin/env bash
#
# hostchk.sh - make various checks on a host to determine if it can use the
# mkiocccentry repo.
#
# NOTE: This is NOT yet finished. More needs to be discussed at
# https://github.com/ioccc-src/mkiocccentry/issues/250 (hostchk rule/script) and at
# https://github.com/ioccc-src/mkiocccentry/issues/248 (make bug-report
# rule/script).
#
# NOTE: This cannot be perfect and it cannot account for everything but along
# with bug_report.sh we hope to account for most things. One can always report
# an issue at the GitHub issues page
# https://github.com/ioccc-src/mkiocccentry/issues (the preferable method as
# others can help as well and in the case of tools - for example txzchk -
# written by someone other than the Judges they can also look at the problem) or
# email the Judges.
#
# This script is being written by:
#
#	@xexyl
# 	https://xexyl.net		Cody Boone Ferguson
#	https://ioccc.xexyl.net
#
# "Because sometimes even the IOCCC Judges need some help." :-)
#
# NOTE: This is a work in progress.
#
# NOTE: This script is called from make all and if it exits non-zero it will
# issue a warning, suggesting you report it via the bug_report.sh script. That
# script itself will call hostchk.sh which will reissue that warning (more than
# once in fact since make all calls hostchk.sh and bug_report.sh calls it via
# make all and running hostchk.sh directly) and each time it will suggest using
# bug_report.sh.
#
# Obviously one needn't run this script a second or third time just because the
# Makefile and bug_report.sh suggest you do! :-) Sorry in advance if this (to
# make use of an American English phrase which I will happily do for a pun for a
# pun not made is a wasted opportunity :-) ) throws you for a loop! :-) But now
# that you're in the loop it shouldn't even matter. :-)
#
# This might seem extra verbose or overkill but we feel that if there's an issue
# with hostchk.sh it really is an issue that will likely prevent a successful
# use of this repo so each time the script fails we report the issue since it
# really is likely that, if the script fails, you will be unable to successfully
# use the mkiocccentry repo to submit a correct IOCCC entry.
#

#

# set up
#

# attempt to fetch system specific paths to the tools we need
#
# get cc path
CC="$(type -P cc 2>/dev/null)"
# It's possible that a path could not be obtained so we set to defaults in this
# case.
#
# We could do it via parameter substitution but since it tries to execute the
# command if for some reason the tool ever works without any args specified it
# would make the script block (if we did it via parameter substitution we would
# still have to redirect stderr to /dev/null). It would look like:
#
#   ${CC:=/usr/bin/cc} 2>/dev/null
#
# make sure CC is set
if [[ -z "$CC" ]]; then
    CC="/usr/bin/cc"
fi

export WORK_DIR="./hostchk"
export HOSTCHK_VERSION="0.3 2022-10-27"
export USAGE="usage: $0 [-h] [-V] [-v level] [-D dbg_level] [-c cc] [-w work_dir] [-f]

    -h			    print help and exit
    -V			    print version and exit
    -v level		    set verbosity level for this script: (def level: 0)
    -D dbg_level	    set verbosity level for tests (def: level: 0)
    -c cc		    path to compiler (def: $CC)
    -w work_dir		    change work directory (def: $WORK_DIR)
    -f			    faster check (def: run slower for better diagnostics)

Exit codes:
     0   all is well
     1   at least one test failed
     2   help mode and version mode exit
     3   invalid command line
 >= 10   internal error

$0 version: $HOSTCHK_VERSION"

export EXIT_CODE=0

# parse args
#
export V_FLAG="0"
export DBG_LEVEL="0"
export F_FLAG=
while getopts :hv:VD:c:w:f flag; do
    case "$flag" in
    h)	echo "$USAGE" 1>&2
	exit 2
	;;
    v)	V_FLAG="$OPTARG";
	;;
    V)	echo "$HOSTCHK_VERSION"
	exit 2
	;;
    D)	DBG_LEVEL="$OPTARG";
	;;
    c)	CC="$OPTARG";
	;;
    w)	WORK_DIR="$OPTARG";
	;;
    f)	F_FLAG="true";
	;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
	exit 3
	;;
    :)	echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
	exit 3
	;;
   *)
	;;
    esac
done

# firewall checks
#
# check for cc
#
if [[ ! -e $CC ]]; then
    echo "$0: ERROR: cc not found: $CC" 1>&2
    exit 10
fi
if [[ ! -f $CC ]]; then
    echo "$0: ERROR: cc not a regular file: $CC" 1>&2
    exit 10
fi
if [[ ! -x $CC ]]; then
    echo "$0: ERROR: cc not executable: $CC" 1>&2
    exit 10
fi

# create an empty working directory
#
rm -rf  "$WORK_DIR"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: rm -rf $WORK_DIR failed, error code: $status" 1>&2
    exit 11
fi
mkdir -p "$WORK_DIR"
status="$?"
if [[ $status -ne 0 ]]; then
    echo "$0: ERROR: mkdir -p $WORK_DIR failed, error code: $status" 1>&2
    exit 11
fi


# set up for compile test
#
export INCLUDE_TEST_SUCCESS="true"
RUN_INCLUDE_TEST="true"
PROG_FILE=$(mktemp -u .hostchk.prog.XXXXXXXXXX)
status="$?"
if [[ $status -ne 0 ]]; then
    EXIT_CODE=41
    echo "$0: ERROR: mktemp -u $PROG_FILE exit code: $status: new exit code: $EXIT_CODE" 1>&2
    RUN_INCLUDE_TEST=
fi

# trap to remove prog and also work directory
trap "rm -rf \$PROG_FILE \$WORK_DIR; exit" 1 2 3 15

# Previously, -f was so fast it did absolutely nothing! :-)
#
# Now, however, fast mode is actually slower! :-( It however can be forgiven,
# perhaps, because it actually does something. :-) Instead of trying to compile
# a program for each required system header file it compiles one source file
# with all of them included at once.
#
if [[ -n "$F_FLAG" ]]; then
    rm -f "$PROG_FILE"
    if [[ $V_FLAG -gt 1 ]]; then
	echo "$0: test compiling test file with all necessary system include files" 1>&2
    fi

    printf "%s\n%s\n" "$(grep -h -o '#include.*<.*>' ./*.c ./*.h|sort -u)" "int main(void) { return 0; }" |
	    "${CC}" -x c - -o "$PROG_FILE"
    # test compile
    #
    status="$?"
    if [[ $status -ne 0 ]]; then
	EXIT_CODE=42
	echo "$0: ERROR: unable to compile test file with all necessary system include files $h: new exit code: $EXIT_CODE" 1>&2
	INCLUDE_TEST_SUCCESS="false"
    elif [[ -s "$PROG_FILE" && -x "$PROG_FILE" ]]; then
	./"$PROG_FILE"
	status="$?"
	if [[ $status -ne 0 ]]; then
	    EXIT_CODE=43
	    echo "$0: ERROR: unable to run executable compiled with all necessary system include files: new exit code: $EXIT_CODE" 1>&2
	    INCLUDE_TEST_SUCCESS="false"
	else
	    if [[ $V_FLAG -gt 1 ]]; then
		echo "$0: Running executable compiled with all necessary system include files: success" 1>&2
	    fi
	fi
    else
	EXIT_CODE=44
	echo "$0: ERROR: unable to form an executable compiled using all necessary system include files: new exit code: $EXIT_CODE" 1>&2
	INCLUDE_TEST_SUCCESS="false"
    fi

    # clean up after compile test
    #
    rm -f "$PROG_FILE"
else
# run include tests one at a time
#
    if [[ -n $RUN_INCLUDE_TEST ]]; then

	# test each required system include file
	#
	while read -r h; do

	    if [[ $V_FLAG -ge 1 ]]; then
		echo "Compiling code with $h" 1>&2
	    fi

	    # form C prog
	    #
	    rm -f "$PROG_FILE"
	    printf "%s\n%s\n" "$h" "int main(void){ return 0; }" | "${CC}" -x c - -o "$PROG_FILE"

	    # test compile
	    #
	    status="$?"
	    if [[ $status -ne 0 ]]; then
		EXIT_CODE=42
		echo "$0: ERROR: unable to compile with $h: new exit code: $EXIT_CODE" 1>&2
		INCLUDE_TEST_SUCCESS="false"
	    elif [[ -s "$PROG_FILE" && -x "$PROG_FILE" ]]; then
		./"$PROG_FILE"
		status="$?"
		if [[ $status -ne 0 ]]; then
		    EXIT_CODE=43
		    echo "$0: ERROR: unable to run executable compiled using: $h: new exit code: $EXIT_CODE" 1>&2
		    INCLUDE_TEST_SUCCESS="false"
		else
		    if [[ $V_FLAG -gt 1 ]]; then
			echo "$0: Running executable compiled using: $h: success"
		    fi
		fi
	    else
		EXIT_CODE=44
		echo "$0: ERROR: unable to form an executable compiled using: $h: new exit code: $EXIT_CODE" 1>&2
		INCLUDE_TEST_SUCCESS="false"
	    fi

	    # clean up after compile test
	    #
	    rm -f "$PROG_FILE"
	done < <(grep -h -o '#include.*<.*>' ./*.c ./*.h|sort -u)
    else
	EXIT_CODE=45
	echo "$0: notice: include test disabled due to test set up error(s): new exit code: $EXIT_CODE" 1>&2
	INCLUDE_TEST_SUCCESS="false"
    fi
fi

# now make and test compile and run some C files

# compile_test	- compile test source file, checking that an executable was
#		  created
#
# usage:
#	compile exit_code source output
#
#	exit_code   - new exit code of rule fails
#	source	    - source file
#	output	    - output file of compiler
#
compile_test() {

    # parse args
    #
    if [[ $# -ne 3 ]]; then
	echo "$0: ERROR: function expects 3 args, found $#" 1>&2
	return 1
    fi
    local CODE="$1"
    local SOURCE="$2"
    local OUTPUT="$3"

    if [[ ! -e "$SOURCE" ]]; then
	echo "$0: ERROR: source code does not exist: $SOURCE: new exit code: $CODE" 1>&2
	EXIT_CODE="$CODE"
	return 1
    fi

    if [[ ! -r "$SOURCE" ]]; then
	echo "$0: ERROR: source file not readable: $SOURCE: new exit code: $CODE" 1>&2
	EXIT_CODE="$CODE"
	return 1
    fi

    $CC -o "$OUTPUT" "$SOURCE"
    if [[ ! -e "$OUTPUT" ]]; then
	echo "$0: ERROR: could not create executable from $SOURCE: new exit code: $CODE" 1>&2
	EXIT_CODE="$CODE"
	return 1
    fi
    if [[ ! -x "$WORK_FILE" ]]; then
	echo "$0: ERROR: could not create executable from $SOURCE: new exit code: $CODE" 1>&2
	EXIT_CODE="$CODE"
    fi

    return 0
}


# pre-errno test: get value of ENOENT
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should print the value of ENOENT" 1>&2
fi
cat <<EOF >"$WORK_DIR/pre-errno.c"
#include <errno.h>
#include <stdio.h>
int
main(void)
{
    printf("%d", ENOENT);

    return 0;
}
EOF
WORK_FILE="$WORK_DIR/pre-errno"
if compile_test 46 "$WORK_DIR/pre-errno.c" "$WORK_FILE"; then
    ENOENT=$("$WORK_FILE")
    if [[ $V_FLAG -gt 1 ]]; then
	echo "$0: got: ENOENT == $ENOENT" 1>&2
    fi
fi
# errno test 0: test printf of errno == ENOENT. If program does not print:
#
#   errno: $ENOENT
#
# where $ENOENT was obtained above then there's an inconsistency with the errno
# value.
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should print: errno: $ENOENT" 1>&2
fi
cat <<EOF >"$WORK_DIR/errno0.c"
#include <errno.h>
#include <stdio.h>
int
main(void)
{
    errno = ENOENT;

    (void) printf("errno: %d\n", errno);
    return 0;
}
EOF
WORK_FILE="$WORK_DIR/errno0"
if compile_test 47 "$WORK_DIR/errno0.c" "$WORK_FILE"; then
    if ! "$WORK_FILE" | grep -q "errno: $ENOENT"; then
	# XXX should we not use -q to grep so we can print the actual value?
	EXIT_CODE=48
	echo "$0: ERROR: inconsistent ENOENT value: did not get: \"errno: $ENOENT\": new exit code: $EXIT_CODE" 1>&2
    elif [[ $V_FLAG -gt 1 ]]; then
	echo "$0: got: errno: $ENOENT" 1>&2
    fi
fi

# errno1: create source file that when compiled should return 1 (errno is set to
# ENOENT so errno == ENOENT should be 1)
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should verify errno == $ENOENT, returning 1" 1>&2
fi
cat<<EOF >"$WORK_DIR/errno1.c"
#include <errno.h>
int
main(void)
{
    errno = ENOENT;

    return errno == ENOENT;
}
EOF
WORK_FILE="$WORK_DIR/errno1"
if compile_test 49 "$WORK_DIR/errno1.c" "$WORK_FILE"; then
    "$WORK_FILE"
    status=$?
    if [[ "$status" -ne 1 ]] ; then
	EXIT_CODE=50
	echo "$0: ERROR: errno != ENOENT: new exit code: $EXIT_CODE" 1>&2
    elif [[ "$V_FLAG" -gt 1 ]]; then
	echo "$0: got return value 1: errno == ENOENT"
    fi
fi

# hello.c: print "Hello, world".
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should print: "Hello, world"" 1>&2
fi
cat <<EOF >"$WORK_DIR/hello.c"
#include <stdio.h>
int
main(void)
{
    (void) printf("Hello, world\n");
    return 0;
}
EOF
WORK_FILE="$WORK_DIR/hello"
if compile_test 51 "$WORK_DIR/hello.c" "$WORK_FILE"; then
    if ! "$WORK_FILE" | grep -q "Hello, world"; then
	EXIT_CODE=52
	echo "$0: ERROR: expected string "Hello, world" not found: new exit code: $EXIT_CODE" 1>&2
    elif [[ "$V_FLAG" -gt 1 ]]; then
	echo "$0: got: "Hello, world"" 1>&2
    fi
fi

# main0: create source file that should return 0
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should simply return 0" 1>&2
fi
cat <<EOF >"$WORK_DIR/main0.c"
int
main(void)
{
    return 0;
}
EOF
WORK_FILE="$WORK_DIR/main0"
if compile_test 52 "$WORK_DIR/main0.c" "$WORK_FILE"; then
    if ! "$WORK_FILE"; then
	EXIT_CODE=53
	echo "$0: ERROR: expected return value 0 not returned: got $?: new exit code: $EXIT_CODE" 1>&2
    elif [[ "$V_FLAG" -gt 1 ]]; then
	echo "$0: got return value 0" 1>&2
    fi
fi


# main1: create source file that should return 1
if [[ $V_FLAG -gt 1 ]]; then
    echo "$0: creating source that should simply return 1" 1>&2
fi
cat <<EOF >"$WORK_DIR/main1.c"
int
main(void)
{
    return 1;
}
EOF
WORK_FILE="$WORK_DIR/main1"
if compile_test 54 "$WORK_DIR/main1.c" "$WORK_FILE"; then
    if "$WORK_FILE"; then
	EXIT_CODE=55
	echo "$0: ERROR: expected return value 1 not returned: got $?: new exit code: $EXIT_CODE" 1>&2
    elif [[ "$V_FLAG" -gt 1 ]]; then
	echo "$0: got return value 1" 1>&2
    fi
fi

# remove work directory
rm -rf "$WORK_DIR"

# All Done!!! -- Jessica Noll, Age 2
#
exit "$EXIT_CODE"
